Integ-test: add run_benchmark fixture to draft a general scaling test framework

hanwen-pcluste opened the pull request:
See commits descriptions for details

Signed-off-by: Hanwen <hanwenli@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Integ-tests: add run_benchmark fixture to draft a general scaling test infrastructure

## A run_benchmarks fixture

A fixture called run_benchmarks will be created in conftest.py. Because the fixture can collect test information from other fixtures, the caller of the fixture only needs to provide two arguments: remote_command_executor and scheduler_commands.

The fixture will run tests specified by benchmarks, conditioned on request.config.getoption("benchmarks"). At the end, the fixture will push the result to CloudWatch metrics.

## A benchmarks fixture

The benchmarks fixture will come from the test config yaml file, similar to the region, scheduler, os, instance fixtures. Therefore, a benchmarks section will be added to the test config YAML file.

## A flag request.config.getoption("benchmarks")

Because we are plugging benchmark tests into existing tests, it is necessary to have a flag to control whether the benchmark tests are executed to avoid always running the benchmarks together with functionality test runs. Benchmarks should be refrained from frequent run to reduce resource consumption.
The --benchmarks flag is provided through test_runner command, e.g. python test_runner.py --key-name "mykey" —key-path /Users/hanwen/.ssh/mykey.pem —parallelism 8 —tests-config configs/develop.yaml --benchmarks.
When scheduling tests on Jenkins, Jenkins Parametrized Scheduler (https://plugins.jenkins.io/parameterized-scheduler/) can be used to exclude benchmarks flag from frequent run and include it in sparce runs.

## Add placement groups to queues when benchmarks are enabled

Placement groups reduce performance noises across different test runs

## Relax timeout of some util functions

When running commands with more nodes, the execution time could be longer

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Use smaller active directories

AWS Managed Microsoft AD (Standard Edition) includes 1 GB of directory object storage. This capacity can support up to 5,000 users or 30,000 directory objects, including users, groups, and computers. AWS Managed Microsoft AD (Enterprise Edition) includes 17 GB of directory object storage, which can support up to 100,000 users or 500,000 objects.

Small Simple AD can have up to 2000 objects, including ~500 users, groups and computers each. Simple AD can have up to 20,000 objects, including ~5000 users, groups and computers each.

Using smaller ADs in the test does not only save cost, but make potential performance issue easier to be discovered.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: move `render_jinja_template` to common file

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: avoid compiling OSU benchmark again if directory already exist

Signed-off-by: Hanwen <hanwenli@amazon.com>

charlesg3 commented on tests/integration-tests/conftest.py:
```diff
@@ -112,8 +114,10 @@ def pytest_addoption(parser):
-    parser.addoption("--benchmarks-max-time", help="set the max waiting time in minutes for benchmarks tests", type=int)
+    parser.addoption("--benchmarks", action="store_true", default=False, help="enable benchmark tests")
+    # parser.addoption("--benchmarks-target-capacity", help="set the target capacity for benchmarks tests", type=int)
+    # parser.addoption("--benchmarks-max-time",
```
Is there a reason these are staying here and commented? If so let's add a comment to how they are useful, otherwise let's remove them.

hanwen-pcluste replied:
I'd remove these. Because demartinofra, the original author, agreed to change the scaling test approach

charlesg3 commented on tests/integration-tests/test_runner.py:
```diff
@@ -437,14 +437,16 @@ def _get_pytest_args(args, regions, log_file, out_dir):  # noqa: C901
     pytest_args.append(f"--key-path={args.key_path}")
     pytest_args.extend(["--stackname-suffix", args.stackname_suffix])
 
+    # if args.benchmarks:
```
Also curious about this block of comments -- if the block is useful explain why it is here / when to uncomment, otherwise let's remove

hanwen-cluster committed:
Integ-tests: rename `num_of_instances` to `num_instances`

Signed-off-by: Hanwen <hanwenli@amazon.com>

charlesg3 commented on tests/integration-tests/tests/common/osu_common.py:
```diff
@@ -103,11 +105,50 @@ def run_osu_benchmarks(
+                        "PacketSize": packet_size,
+                    }
+                )
+                metric_data.append(
+                    {
+                        "MetricName": "Latency",
+                        "Dimensions": [{"Name": name, "Value": str(value)} for name, value in dimensions_copy.items()],
+                        "Value": int(latency),
+                        "Unit": "Microseconds",
+                    }
+                )
+                yield metric_data
```
Doesn't this continually yield longer and longer lists rather than a single list of metric data?

hanwen-pcluste replied:
Fixed, thank you!
Interestingly, the CloudWatch Metrics looked perfect.

charlesg3 reviewed (changes requested):
A few minor comments inlined.

charlesg3 commented on tests/integration-tests/utils.py:
```diff
@@ -12,6 +12,7 @@
 import json
 import logging
 import os
+import os as os_lib
```
Is there any reason to import `os` twice? If not let's import it a single time and update the code to reference the single import.

hanwen-pcluste replied:
Done

hanwen-cluster committed:
Integ-tests: Update README.md

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Push job wait time to CloudWatch

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Make slots per instance configurable in performance test

This is useful when hyperthreading is disabled

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Enable performance test for test disable hyperthreading

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Use larger max count to support performance tests if not specified explicitly.

The default max count is 10, which is too small for performance tests.

This change to the cluster config file is strictly limited:
1. The change is only applied when `--benchmarks` flag is used and the specific test does have benchmarks to run in the test definition file.
2. If the max count is explicitly specified in the cluster config, we don't change the max count

This is useful when hyperthreading is disabled

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Use smaller max count for AD integration test

Specifying smaller max count safeguards the cost of the tests.

The 3000 max count was put in place for manual testing during the early stage of implementation.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Integ-tests: Enable performance test for test ad integration

Signed-off-by: Hanwen <hanwenli@amazon.com>

charlesg3 commented on tests/integration-tests/README.md:
````diff
@@ -771,13 +755,107 @@ def vpc(cfn_stacks_factory):
+              other_benchmarks:
+                other_benchmarks_group: ["other_allreduce", "other_alltoall"]
+```
+2. The `run_benchmarks` fixture runs benchmarks for different `mpi_variants`, `num_instances` and push the result to CloudWatch with the correct namespace. The detailed execution of the OSU benchmarks is in `osu_common.py`, decoupled from the `run_benchmarks` fixture. So in the future, the `run_benchmarks` fixture will look like:
````
and push*es* the result to CloudWatch...

charlesg3 commented on tests/integration-tests/tests/common/osu_common.py:
```diff
@@ -91,23 +92,78 @@ def run_osu_benchmarks(
+                # as an indicator of how fast the compute nodes were launched.
+                metric_data.append(
+                    {
+                        "MetricName": "SignificantJobWaitTime",
```
`JobWaitTime` to be more concise with the MetricName parameter (?)

hanwen-pcluste committed:
Merge branch 'develop' into scaling_test
